Zend/zend_closures.c

```diff
@@ -381,7 +381,35 @@ static ZEND_COLD zend_function *zend_closure_get_constructor(zend_object *object
 static int zend_closure_compare(zval *o1, zval *o2) /* {{{ */
 {
 	ZEND_COMPARE_OBJECTS_FALLBACK(o1, o2);
-	return Z_OBJ_P(o1) != Z_OBJ_P(o2);
+
+	zend_closure *lhs = (zend_closure*) Z_OBJ_P(o1);
+	zend_closure *rhs = (zend_closure*) Z_OBJ_P(o2);
+
+	if (!((lhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE) && (rhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE))) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (Z_OBJ(lhs->this_ptr) != Z_OBJ(rhs->this_ptr)) {
```

Comment:
Is it safe to directly compare these? I'm not sure we initialize Z_OBJ if type type is UNDEF. I would have expected something along these lines here:
```
if (Z_TYPE(lhs->this_ptr) != Z_TYPE(rhs->this_ptr)) {
    return ZEND_UNCOMPARABLE;
}
if (Z_TYPE(lhs->this_ptr) == IS_OBJECT && Z_OBJ(lhs->this_ptr) != Z_OBJ(rhs->this_ptr)) {
    return ZEND_UNCOMPARABLE;
}
```

Reply:
Ah, actually no it's not ... fixed in 67c0361ab9a60aa87339717fe4c26c34cbbcb976

Reply:
Well ... it looked safe because the closure is memset, but since we allow access to this_ptr there's no guarantee, so this is probably the safer thing to do anyway ...